src/P3_particle_properties.jl

```diff
@@ -1,333 +1,498 @@
-    else
-        @assert ρ_r > FT(0)   # rime density must be positive ...
-        @assert ρ_r <= p3.ρ_l # ... and as a bulk ice density can't exceed the density of water
-
-        P3_problem(ρ_d) =
-            ρ_d - ρ_d_helper(
-                p3,
-                D_cr_helper(p3, F_rim, ρ_g_helper(ρ_r, F_rim, ρ_d)),
-                D_gr_helper(p3, ρ_g_helper(ρ_r, F_rim, ρ_d)),
-            )
-
-        ρ_d =
-            RS.find_zero(
-                P3_problem,
-                RS.SecantMethod(FT(0), FT(1000)),
-                RS.CompactSolution(),
-            ).root
-        ρ_g = ρ_g_helper(ρ_r, F_rim, ρ_d)
-
-        return (;
-            D_cr = D_cr_helper(p3, F_rim, ρ_g),
-            D_gr = D_gr_helper(p3, ρ_g),
-            ρ_g,
-            ρ_d,
-        )
-    end
-end
+Return the size of equal mass for graupel and partially rimed ice [meters]
 
+See Eq. 14 in [MorrisonMilbrandt2015](@cite).
 """
-    p3_F_liq_average(F_liq, X_ice, X_liq)
+get_D_cr(mass::MassPowerLaw, ρ_g, F_rim) =
+    get_threshold(mass, ρ_g * (1 - F_rim))
 
- - F_liq - liquid fraction (L_liq / L_p3_tot)
- - X_ice - ice core parameterization (i.e. mass, etc)
- - X_liq - liquid part parameterization
+"""
+    weighted_average(f_a, a, b)
 
-Returns the liquid fraction weighted average of X_ice and X_liq.
+Return the weighted average of `a` and `b` with fraction `f_a`,
+    
+```math
+f_a ⋅ a + (1 - f_a) ⋅ b
+```
 """
-function p3_F_liq_average(F_liq::FT, X_ice::FT, X_liq::FT) where {FT}
-    return (1 - F_liq) * X_ice + F_liq * X_liq
+function weighted_average(f_a, a, b)
+    return f_a * a + (1 - f_a) * b
 end
 
 """
-    p3_density(p3, D, F_rim, th)
-
-- p3 - a struct with P3 parameters
-- D - maximum particle dimension [m]
-- F_rim - rime mass fraction (L_rim / L_ice) [-]
-- th - P3 scheme thresholds() output tuple (D_cr, D_gr, ρ_g, ρ_d)
-
-Returns the density of a particle based on where it falls in the particle-size-based properties
-regime. Following Morrison and Milbrandt (2015), the density of nonspherical particles is assumed to
-be the particle mass divided by the volume of a sphere with the same D.
-Needed for aspect ratio calculation, so we assume zero liquid fraction.
-"""
-function p3_density(p3::PSP3, D::FT, F_rim::FT, th) where {FT}
-    if D_th_helper(p3) > D
-        # small spherical ice
-        return p3.ρ_i
-    elseif F_rim == 0
-        # large nonspherical unrimed ice
-        return (6 * α_va_si(p3)) / FT(π) * D^(p3.β_va - 3)
-    elseif th.D_gr > D >= D_th_helper(p3)
-        # dense nonspherical ice
-        return (6 * α_va_si(p3)) / FT(π) * D^(p3.β_va - 3)
-    elseif th.D_cr > D >= th.D_gr
-        # graupel
-        return th.ρ_g
-    else #elseif D >= th.D_cr
-        # partially rimed ice
-        return (6 * α_va_si(p3)) / (FT(π) * (1 - F_rim)) * D^(p3.β_va - 3)
+    volume_sphere(D)
+
+Calculate the volume of a sphere with diameter D.
+
+```math
+V = \\frac{4}{3} π D^3
+```
+"""
+volume_sphere(D) = D^3 * π / 6
```

Comment:
```suggestion
volume_sphere(D) = D^3 * FT(π) / 6
```

Reply:
How about:
```julia
volume_sphere(D::FT)::FT where {FT} = D^3 * π / 6
```
this ensures the return type of the function is `FT`.